File: draft_design_pattern/mediator.ts
export {};

interface ChatRoomMediator {
  showMessage(user: User, message: string): void;
}

class ChatRoom implements ChatRoomMediator {
  showMessage(user: User, message: string): void {
    const d = new Date();
    const y = d.getFullYear();
    const m = d.getMonth() + 1;
    const day = d.getDate();
    const ho = d.getHours();
    const mi = d.getMinutes();
    const sender = user.getName();
    console.log(
      `${y}-${m}-${day} ${ho}:${mi}[${sender}]:${message}`,
    );
  }
}

class User {
  constructor(public name: string, public chatMediator: ChatRoomMediator) {}
  getName(): string {
    return this.name;
  }
  send(message: string): void {
    this.chatMediator.showMessage(this, message);
  }
}

const mediator = new ChatRoom();

const john = new User("John Doe", mediator);
const jane = new User("Jane Doe", mediator);

john.send("こんちは！");
jane.send("よう！");

File: draft_design_pattern/mediator.nim
import strformat
import times


type
  ChatRoomMediator = ref object of RootObj
  ChatRoom = ref object of ChatRoomMediator

  User = ref object
    name: string
    chatMediator: ChatRoomMediator


# プロトタイプ宣言
proc getName(self: User): string


method showMessage(self: ChatRoomMediator, user: User, message: string) {.base,
    locks: "unknown".} =
  discard


method showMessage(self: ChatRoom, user: User, message: string) {.
    locks: "unknown".} =
  let date = now().format("yyyy-MM-dd(ddd)HH:mm")
  let sender = user.getName()
  echo fmt"{date}[{sender}]{message}"


proc getName(self: User): string =
  self.name

proc send(self: User, message: string) =
  self.chatMediator.showMessage(self, message)


proc main() =
  let mediator = ChatRoom()

  let john = User(name: "John Doe", chatMediator: mediator)
  let jane = User(name: "Jane Doe", chatMediator: mediator)

  john.send("こんちは！")
  jane.send("よう！")

main()

